dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageProblemCauseAnnotator.java

```diff
@@ -35,74 +39,100 @@ private LinkageProblemCauseAnnotator() {}
+    if (subtreeResult == null) {
+      // Resolves the dependency graph with the source artifact at the root.
+      subtreeResult = classPathBuilder.resolveWithMaven(sourceArtifact);
+      artifactResolutionCache.put(sourceArtifact, subtreeResult);
+    }
 
-      Symbol symbol = linkageProblem.getSymbol();
-      ClassPathEntry entryInSubtree = subtreeResult.findEntryBySymbol(symbol);
-      if (entryInSubtree == null) {
+    Symbol symbol = linkageProblem.getSymbol();
+    ClassPathEntry entryInSubtree = subtreeResult.findEntryBySymbol(symbol);
+    if (entryInSubtree == null) {
+      linkageProblem.setCause(UnknownCause.getInstance());
+    } else {
+      Artifact artifactInSubtree = entryInSubtree.getArtifact();
+      ImmutableList<DependencyPath> dependencyPathsToSource =
+          rootResult.getDependencyPaths(sourceEntry);
+      if (dependencyPathsToSource.isEmpty()) {
+        // When an artifact is excluded, it's possible to have no dependency path to sourceEntry.
         linkageProblem.setCause(UnknownCause.getInstance());
-      } else {
-        Artifact artifactInSubtree = entryInSubtree.getArtifact();
-        DependencyPath pathToSourceEntry = rootResult.getDependencyPaths(sourceEntry).get(0);
-        DependencyPath pathFromSourceEntryToUnselectedEntry =
-            subtreeResult.getDependencyPaths(entryInSubtree).get(0);
-        DependencyPath pathToUnselectedEntry =
-            pathToSourceEntry.concat(pathFromSourceEntryToUnselectedEntry);
+        return;
+      }
+      DependencyPath pathToSourceEntry = dependencyPathsToSource.get(0);
+      DependencyPath pathFromSourceEntryToUnselectedEntry =
+          subtreeResult.getDependencyPaths(entryInSubtree).get(0);
+      DependencyPath pathToUnselectedEntry =
+          pathToSourceEntry.concat(pathFromSourceEntryToUnselectedEntry);
 
-        ClassPathEntry selectedEntry =
-            rootResult.findEntryById(
-                artifactInSubtree.getGroupId(), artifactInSubtree.getArtifactId());
-        if (selectedEntry != null) {
-          Artifact selectedArtifact = selectedEntry.getArtifact();
-          if (!selectedArtifact.getVersion().equals(artifactInSubtree.getVersion())) {
-            // Different version of that artifact is selected in rootResult
-            linkageProblem.setCause(
-                new DependencyConflict(
-                    linkageProblem,
-                    rootResult.getDependencyPaths(selectedEntry).get(0),
-                    pathToUnselectedEntry));
-          } else {
-            // A linkage error was already there when sourceArtifact was built.
+      ClassPathEntry selectedEntry =
+          rootResult.findEntryById(
+              artifactInSubtree.getGroupId(), artifactInSubtree.getArtifactId());
+      if (selectedEntry != null) {
+        Artifact selectedArtifact = selectedEntry.getArtifact();
+        if (!selectedArtifact.getVersion().equals(artifactInSubtree.getVersion())) {
```

Comment:
You could probably combine the guard clause you created here with the other error condition:

```
ClassPathEntry selectedEntry =
    rootResult.findEntryById(
        artifactInSubtree.getGroupId(), artifactInSubtree.getArtifactId());

if (selectedEntry != null) {
  Artifact selectedArtifact = selectedEntry.getArtifact();
  ImmutableList<DependencyPath> pathToSelectedArtifact =
      rootResult.getDependencyPaths(selectedEntry);

  if (pathToSelectedArtifact.isEmpty() || !selectedArtifact.getVersion().equals(artifactInSubtree.getVersion()) {
    linkageProblem.setCause(UnknownCause.getInstance());
    return;
  }
  
  linkageProblem.setCause(
      new DependencyConflict(
          linkageProblem, pathToSelectedArtifact.get(0), pathToUnselectedEntry));
} else {
...
}
```

Reply:
It may be logically the same but the current if-statement delivers better meaning.

Reply:
:+1: